### PR TITLE
feat: Pick defined values from object

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Introduces a type-safe object picker, inspired by [`removeFlaseyValues`][] in Frontend.

[`removeFlaseyValues`]: https://github.com/guardian/frontend/blob/035b8ff00fdd2740e33d4d1ab65da5fb0bb78deb/static/src/javascripts/projects/commercial/modules/header-bidding/utils.ts#L48

## Why?

Code Golfing is fun 🏌️

### CODE REVIEW

File: src/lib/pick-defined-values.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	? never
+	: [''] extends T
+	? never
+	: T extends boolean
+	? never
+	: T extends NonNullable<T>
+	? K
+	: never;
+type DefinedKeys<T> = { [K in keyof T]-?: ValidTargeting<T[K], K> }[keyof T];
+type ObjectWithDefinedValues<T> = Pick<T, DefinedKeys<T>>;
```
Comment (arelra) [inline src/lib/pick-defined-values.ts]: Wow 😎 

5x3bzr

File: src/lib/pick-defined-values.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+const isValidTargeting = (value: unknown): value is string | string[] => {
+	if (isString(value) && value !== '') return true;
+	if (
+		Array.isArray(value) &&
+		value.filter(isString).filter(Boolean).length > 0
+	)
+		return true;
```
Comment (arelra) [inline src/lib/pick-defined-values.ts]: 'Dangling' returns scare me a bit.

Also can we improve the readability a bit and create named variables for the boolean conditions?

File: src/lib/pick-defined-values.ts (hunk 1 above)
Comment (mxdvl) [reply src/lib/pick-defined-values.ts]: 🤣  Yeah I explored these ideas in 10% time… it’s definitely right on the edge of usefulness.

Comment (ashishpuliyel) [reply src/lib/pick-defined-values.ts]: wow is right.

**Review by ashishpuliyel — APPROVED** [review]

**Pull request merged**